source/RenderBuffer.cpp

```diff
@@ -0,0 +1,326 @@
+	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
+
+	// Generate the texture.
+	glGenTextures(1, &texid);
+	glBindTexture(GL_TEXTURE_2D, texid);
+
+	// Use nearest pixel and no wrapping.
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+	// Attach a blank image to the texture.
+	const Point scaledSize = size * Screen::Zoom() / 100;
+	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, scaledSize.X(), scaledSize.Y(), 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
+
+	// Attach the texture to the frame buffer.
+	glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, texid, 0);
+	GLenum draw_buffers[] = {GL_COLOR_ATTACHMENT0};
+	glDrawBuffers(1, draw_buffers);
+
+	if(glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE)
+		Logger::LogError("Failed to initialize framebuffer for RenderBuffer");
+
+
+	glBindTexture(GL_TEXTURE_2D, 0);
+	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+	// Default to the current viewport size at the time of construction.
+	glGetIntegerv(GL_VIEWPORT, lastViewport);
+}
+
+
+
+// Destructor. Frees the texture and renderbuffers.
+RenderBuffer::~RenderBuffer()
+{
+	glDeleteTextures(1, &texid);
+	glDeleteFramebuffers(1, &framebuffer);
+}
+
+
+
+// Turn this buffer on as a render target. The render target is restored if
+// the Activation object goes out of scope.
+RenderBuffer::RenderTargetGuard RenderBuffer::SetTarget()
+{
+	// NOTE: These glGets can cause an unwanted state synchronization that can
+	//       cause performance problems. The only real reason we might want this
+	//       is if we are nesting render buffers. If only one framebuffer is
+	//       enabled at a time, then we can just reset the buffer to 0 when we
+	//       are done.
+	glGetIntegerv(GL_FRAMEBUFFER_BINDING, reinterpret_cast<int*>(&lastFramebuffer));
+	glGetIntegerv(GL_VIEWPORT, lastViewport);
+	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
+
+	const Point scaledSize = size * Screen::Zoom() / 100;
+	glViewport(0, 0, scaledSize.X(), scaledSize.Y());
+
+	static const float CLEAR[] = {0, 0, 0, 0};
+	glClearBufferfv(GL_COLOR, 0, CLEAR);
+
+	return RenderTargetGuard(*this, size.X(), size.Y());
+}
+
+
+
+// Reset the render target and viewport to the original settings.
+void RenderBuffer::Deactivate()
+{
+	// Restore the old settings.
+	glViewport(lastViewport[0], lastViewport[1], lastViewport[2], lastViewport[3]);
+	glBindFramebuffer(GL_FRAMEBUFFER, lastFramebuffer);
+}
+
+
+
+void RenderBuffer::Draw(const Point &position)
+{
+	Draw(position, size);
+}
+
+
+
+// Draw the contents of this buffer at the specified position.
+void RenderBuffer::Draw(const Point &position, const Point &clipsize, const Point &srcposition)
+{
+	glUseProgram(shader.Object());
+	glBindVertexArray(vao);
+
+	glUniform1i(texI, 0);
```

Comment:
this is not really needed but leave it I guess
~the minimum lines in me is screaming, ignore it~

Reply:
I assume you are talking about the `glUniform1i(texI, 0);`. That part of the code I copy-pasted from the sprite shader (see `SpriteShader::Add()`), and I presume it isn't needed there either. 

I'm still fuzzy about which variables the shader will auto-bind, vs which ones need explicitly bound, so let me know if I made any other mistakes like this.